### PR TITLE
Jepsen nemesis and lock fix 1st attempt

### PR DESCRIPTION
**Pull request by gmaretic**
**Goals (and why)**:
Add service failure nemesis and enable lock tests.

**Implementation Description (bullets)**:
Mostly based of @harrybiddle 's code

**Concerns (what feedback would you like?)**:
Not familiar with clojure, so may have done something dubious. Works locally, but have yet to do another thorough pass.

**Where should we start reviewing?**:
lock.clj

**Priority (whenever / two weeks / yesterday)**:
Ideally this week

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2011)

### CODE REVIEW

File: atlasdb-jepsen-tests/src/jepsen/atlasdb/lock.clj [hunk 1]
```diff
@@ -91,18 +92,23 @@
...
+    :nemesis nem
     :generator (->> generator
-                    (gen/stagger 0.1)
-                    (gen/clients)
-                    (gen/time-limit 30))
+                 (gen/stagger 0.05)
+                 (gen/nemesis
+                   (gen/seq (cycle [(gen/sleep 5)
+                                    {:type :info, :f :start}
+                                    (gen/sleep 15)
```
Comment (harrybiddle) [inline atlasdb-jepsen-tests/src/jepsen/atlasdb/lock.clj]: Is 15 seconds appropriate for all the nemeses? How long does the server take to start up?

File: atlasdb-jepsen-tests/src/jepsen/atlasdb/timelock.clj [hunk 2]
```diff
@@ -2,7 +2,19 @@
...
+  (:import com.palantir.atlasdb.http.TimelockUtils))
+
+(defn start!
+  "Starts timelock"
+  [node]
+  (c/su
+    (info node "Starting timelock server")
+    (c/exec :env "JAVA_HOME=/usr/lib/jvm/java-8-oracle" "/timelock-server/service/bin/init.sh" "start")
+    (info node "Waiting until timelock node is ready")
+    (TimelockUtils/waitUntilHostReady (name node))
```
Comment (harrybiddle) [inline atlasdb-jepsen-tests/src/jepsen/atlasdb/timelock.clj]: I don't think we should have this `start!` function in. We explicitly changed the test to _not_ wait until the server had started, and instead already serve requests while it is starting up, to make this more real-world.

File: atlasdb-jepsen-tests/src/jepsen/atlasdb/timestamp.clj [hunk 3]
```diff
@@ -58,18 +58,18 @@
...
-                    (gen/seq (cycle [(gen/sleep 5)
-                                     {:type :info, :f :start}
-                                     (gen/sleep 85)
-                                     {:type :info, :f :stop}])))
-                    (gen/time-limit 360))
+                 (gen/stagger 0.05)
+                 (gen/nemesis
+                   (gen/seq (cycle [(gen/sleep 5)
+                                    {:type :info, :f :start}
+                                    (gen/sleep 15)
```
Comment (harrybiddle) [inline atlasdb-jepsen-tests/src/jepsen/atlasdb/timestamp.clj]: why did you change from 85 to 15 seconds?

File: atlasdb-jepsen-tests/src/jepsen/atlasdb/timestamp.clj [hunk 4]
```diff
@@ -58,18 +58,18 @@
...
+                                    {:type :info, :f :stop}])))
+                 (gen/time-limit 180))
```
Comment (harrybiddle) [inline atlasdb-jepsen-tests/src/jepsen/atlasdb/timestamp.clj]: any reason the time-limit has gone from 360 to 180?

File: atlasdb-jepsen-tests/src/main/java/com/palantir/atlasdb/jepsen/events/InfoEvent.java [hunk 5]
```diff
@@ -40,7 +40,7 @@
 
     public abstract long time();
 
-    public abstract Optional<String> value();
+    public abstract Optional<Object> value();
```
Comment (harrybiddle) [inline atlasdb-jepsen-tests/src/main/java/com/palantir/atlasdb/jepsen/events/InfoEvent.java]: I'm not sure about this change. I think the reason this was put in was because the `:info` event for the crashing nemesis wasn't castable as a string. Could you confirm this is the case? Unsure if having `Object` here is going to cause problems elsewhere, although if the tests pass then it should be OK.

File: atlasdb-jepsen-tests/src/jepsen/atlasdb/lock.clj [hunk 6]
```diff
@@ -91,18 +92,23 @@
...
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;; Defining the Jepsen test
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 (defn lock-test
-  []
+  [nem]
   (assoc tests/noop-test
     :client (create-client nil nil nil)
+    :nemesis nem
```
Comment (harrybiddle) [inline atlasdb-jepsen-tests/src/jepsen/atlasdb/lock.clj]: 👍  great idea to pass this in!

File: atlasdb-jepsen-tests/test/jepsen/atlasdb_test.clj [hunk 7]
```diff
@@ -1,14 +1,22 @@
...
 
-(deftest timestamp-test
-   (is (:valid? (:results (jepsen/run! (timestamp/timestamp-test))))))
+(deftest timestamp-test-crash
+  (is (:valid? (:results (jepsen/run! (timestamp/timestamp-test timelock/crash-nemesis))))))
 
-(deftest lock-test
-   (is (:valid? (:results (jepsen/run! (lock/lock-test))))))
+(deftest timestamp-test-partition
+  (is (:valid? (:results (jepsen/run! (timestamp/timestamp-test (nemesis/partition-random-halves)))))))
```
Comment (harrybiddle) [inline atlasdb-jepsen-tests/test/jepsen/atlasdb_test.clj]: why the extra brackets here? `(nemesis/partition-random-halves)` is going to call the `nemesis/partition-random-halves` function, but I think all we want to do is to pass it in (i.e. as a functor/function object). Are you sure this is working?

File: atlasdb-jepsen-tests/src/jepsen/atlasdb/timelock.clj [hunk 8]
```diff
@@ -34,3 +45,34 @@
...
+    (min 2)
+    (take (shuffle xs))))
+
+(def crash-nemesis
+  "A nemesis that crashes a random subset of nodes."
+  (nemesis/node-start-stopper
+    mostly-small-nonempty-subset-at-most-two
+    (fn start [test node] (c/su
+                            (c/exec :killall :-9 :java))
+      ;                            (c/exec :rm :-r "/atlasdb-timelock-server/var/data/paxos"))
```
Comment (harrybiddle) [inline atlasdb-jepsen-tests/src/jepsen/atlasdb/timelock.clj]: Suggest you put a comment here explaining why you've removed this line

**Comment by harrybiddle** [comment]
We should definitely ensure that this passes the Jenkins CI job before merging - @gmaretic have you done that already? We'd have to re-configure the job to point at your branch.

File: atlasdb-jepsen-tests/src/main/java/com/palantir/atlasdb/jepsen/events/InfoEvent.java (hunk 5 above)
Comment (gmaretic) [inline atlasdb-jepsen-tests/src/main/java/com/palantir/atlasdb/jepsen/events/InfoEvent.java]: That is correct, I was getting an exception when trying to cast it to String. I think in general the way that the crashing nemesis is written is not exactly how nemeses should look like, in the sense that it does work, but cannot be composed with the compose nemesis to combine it with others. But that "fix" is definitely a stretch goal as the functionality is what we care about.

File: atlasdb-jepsen-tests/test/jepsen/atlasdb_test.clj (hunk 7 above)
Comment (gmaretic) [inline atlasdb-jepsen-tests/test/jepsen/atlasdb_test.clj]: It doesn't work **without** the extra parentheses. Clojure is magic...

File: atlasdb-jepsen-tests/src/jepsen/atlasdb/timelock.clj (hunk 2 above)
Comment (gmaretic) [inline atlasdb-jepsen-tests/src/jepsen/atlasdb/timelock.clj]: Fair enough, will change back

File: atlasdb-jepsen-tests/src/jepsen/atlasdb/lock.clj (hunk 1 above)
Comment (gmaretic) [inline atlasdb-jepsen-tests/src/jepsen/atlasdb/lock.clj]: This snuck in from testing. I'll put it to something more reasonable

File: atlasdb-jepsen-tests/src/jepsen/atlasdb/timestamp.clj (hunk 4 above)
Comment (gmaretic) [inline atlasdb-jepsen-tests/src/jepsen/atlasdb/timestamp.clj]: We do run each test twice now, but maybe we can leave it at the old time limit

File: atlasdb-jepsen-tests/src/jepsen/atlasdb/timestamp.clj (hunk 3 above)
Comment (gmaretic) [inline atlasdb-jepsen-tests/src/jepsen/atlasdb/timestamp.clj]: This snuck in from testing

File: atlasdb-jepsen-tests/src/jepsen/atlasdb/timestamp.clj (hunk 4 above)
Comment (harrybiddle) [inline atlasdb-jepsen-tests/src/jepsen/atlasdb/timestamp.clj]: But it's gone from running a partition_random_halves for 360 seconds, to running partition_random_halves for 180 and then crash for 180, right? I'd keep it at 360, I don't think we need to worry about the run time so much.

File: atlasdb-jepsen-tests/src/main/java/com/palantir/atlasdb/jepsen/events/InfoEvent.java (hunk 5 above)
Comment (harrybiddle) [inline atlasdb-jepsen-tests/src/main/java/com/palantir/atlasdb/jepsen/events/InfoEvent.java]: Definitely agree - getting in to the weeds of how the nemeses work is _not_ worth it. Cool, let's leave it as `Object` then :)

File: atlasdb-jepsen-tests/test/jepsen/atlasdb_test.clj (hunk 7 above)
Comment (harrybiddle) [inline atlasdb-jepsen-tests/test/jepsen/atlasdb_test.clj]: Ohh, I see that it used to be `(nemesis/partition-random-halves)` before the refactor. Fair enough!

**Comment by gmaretic** [comment]
I have only tested it locally so far, but definitely should try it on circle before merging

File: atlasdb-jepsen-tests/build.gradle [hunk 9]
```diff
@@ -31,6 +31,7 @@ dependencies {
     compile group: 'com.fasterxml.jackson.core', name: 'jackson-databind'
     compile group: 'com.fasterxml.jackson.datatype', name: 'jackson-datatype-jdk8'
     compile group: 'com.google.guava', name: 'guava'
+    compile group: 'com.jayway.awaitility', name: 'awaitility'
```
Comment (harrybiddle) [inline atlasdb-jepsen-tests/build.gradle]: We should be able to remove this now, right?

**Comment by harrybiddle** [comment]
Unfortunately I couldn't get the tests to run on CircleCI, because it didn't work without privileged containers :(. Jenkins was the only option. We should _really_ improve this situation.

File: atlasdb-jepsen-tests/src/jepsen/atlasdb/timelock.clj [hunk 10]
```diff
@@ -34,3 +41,36 @@
...
+(defn mostly-small-nonempty-subset-at-most-two
+  "Returns a subset of the given collection, with a logarithmically decreasing
+  probability of selecting more elements. Always selects at least one element.
+      (->> #(mostly-small-nonempty-subset [1 2 3 4 5])
+           repeatedly
+           (map count)
+           (take 10000)
+           frequencies
+           sort)
+      ; => ([1 3824] [2 2340] [3 1595] [4 1266] [5 975])"
```
Comment (harrybiddle) [inline atlasdb-jepsen-tests/src/jepsen/atlasdb/timelock.clj]: +1! Good flag, I think we could remove the `min 2` step

File: atlasdb-jepsen-tests/src/jepsen/atlasdb/lock.clj [hunk 11]
```diff
@@ -2,13 +2,14 @@
   (:require [jepsen.atlasdb.timelock :as timelock]
             [jepsen.checker :as checker]
             [jepsen.client :as client]
+            [jepsen.control :as c]
```
Comment (harrybiddle) [inline atlasdb-jepsen-tests/src/jepsen/atlasdb/lock.clj]: Agree - good catch

**Comment by gmaretic** [comment]
This addresses https://github.com/palantir/atlasdb/issues/1431

**Comment by gmaretic** [comment]
@harrybiddle Are we good to merge?

**Comment by harrybiddle** [comment]
Can you remove the awaitility dependency that was added? I'd also wouldn't include any release notes, since the PR doesn't make any changes to the AtlasDB product.

Otherwise looks fine to me, will approve so I'm not blocking the PR.
